warehouse/macaroons/caveats.py

```diff
@@ -24,9 +25,16 @@ class InvalidMacaroonError(Exception):
 class Caveat:
     def __init__(self, verifier):
         self.verifier = verifier
+        self.failure_reason = None
+
+    def _fail(self, reason):
+        # TODO: Surface this failure reason to the user.
+        # See: https://github.com/pypa/warehouse/issues/9018
```

Comment:
This change will make #9018 much easier to implement -- we just need to plumb the `failure_reason` through the verifier and service layers.